### PR TITLE
[RN Mobile] RichText: Return early when content has changed in onSelectionChangeFromAztec

### PR DESCRIPTION
**Pull request by kbrown9**
## What?
If the component's content has changed, return early in the `RichText::onSelectionChangedFromAztec` method.

## Why?

There have been reports of slow performance in the RichText component on older and lower-end devices. One potential cause of the slow performance is the amount of work that happens with each keypress in the `RichText::onSelectionChangeFromAztec` method.

Issue: https://github.com/wordpress-mobile/gutenberg-mobile/issues/2594

## How?

It seems unnecessary to do anything in the `RichText::onSelectionChangeFromAztec` method while the user is currently typing quickly. It appears that the `RichText:onChangeFromAztec` method is only called at the beginning and end of a typing spree. Therefore, a potential improvement is to only update the selection when `RichText::onChangeFromAztec` has been called. We can determine whether `RichText::onChangeFromAztec` has been called by comparing the event's content to the component's current content and checking the last event type. If they are different and the last event type was 'selection change', `RichText::onChangeFromAztec` was not called. (If it had been called, it would have set the component's content to the event's content [here](https://github.com/WordPress/gutenberg/blob/trunk/packages/rich-text/src/component/index.native.js#L341)). 

I expect this change to significantly decrease the lag experienced when creating a new block after typing. I also expect this change to decrease the lag experienced when a block using the RichText component expands to create a new line after a significant amount of typing.

## Testing Instructions
The [writing flows](https://github.com/wordpress-mobile/test-cases/tree/trunk/test-cases/gutenberg/writing-flow) will need to be tested thoroughly to verify that this change doesn't cause any regressions.

## Screenshots or screencast 
n/a

### CODE REVIEW

**Comment by SiobhyB** [comment]
👋 @kbrown9, thank you again for this PR. 🙇‍♀️ To help with testing, I've created installable builds for Android [here](https://github.com/wordpress-mobile/WordPress-Android/pull/16742#issuecomment-1155203187) and iOS [here](https://github.com/wordpress-mobile/WordPress-iOS/pull/18873#issuecomment-1155254454). I'll use those builds to go through the writing flow tests for both platforms and aim to finish my review/testing by EOD tomorrow.

**Comment by kbrown9** [comment]
> 👋 @kbrown9, thank you again for this PR. 🙇‍♀️ To help with testing, I've created installable builds for Android [here](https://github.com/wordpress-mobile/WordPress-Android/pull/16742#issuecomment-1155203187) and iOS [here](https://github.com/wordpress-mobile/WordPress-iOS/pull/18873#issuecomment-1155254454). I'll use those builds to go through the writing flow tests for both platforms and aim to finish my review/testing by EOD tomorrow.

Thank you for creating those builds! I had no idea where to start with that. I'll plan to install the Android build on my devices for testing. Unfortunately, I don't have any iOS devices to use for testing.

As you can see from the code changes, I added a check for the last event type. The mobile unit tests caught a problem when changing the color of a text selection, and checking the last event type solved that problem.

**Comment by SiobhyB** [comment]
> Thank you creating those builds! I had no idea where to start with that.

No worries! Just in case you're interested in playing more with Mobile Gutenberg in the future, we have some documentation (that's a bit hidden away) to help with creating installable builds for Android [here](https://github.com/wordpress-mobile/gutenberg-mobile/blob/trunk/WPAndroid%20Integration.md) and iOS [here](https://github.com/wordpress-mobile/gutenberg-mobile/blob/trunk/WPiOS%20Integration.md).

> I'll plan to install the Android build on my devices for testing. Unfortunately, I don't have any iOS devices to use for testing.

Perfect, if you're able to go through the checklist for Android then I'll go through it for iOS.

<hr />

### 🟢 Writing Flow checklist (iOS)

**Build:** [pr18873-f758713](https://github.com/wordpress-mobile/WordPress-iOS/pull/18873#issuecomment-1155254454)
**Device:** iPhone XR (iOS 15.4.1)

#### General
- [x] [TC001](https://github.com/wordpress-mobile/test-cases/blob/trunk/test-cases/gutenberg/writing-flow/copy-paste.md#tc001) - Paste formatted text copied from website
- [x] [TC001](https://github.com/wordpress-mobile/test-cases/blob/trunk/test-cases/gutenberg/writing-flow/multiline-components.md#tc001) - Multiline components
  - [x] Quote block
  - [x] Verse block
  - [x] Preformatted block
  - [ ] Code block (DEV only)
  - [x] Pullquote block
#### Rich Text Format
- [x] [TC001](https://github.com/wordpress-mobile/test-cases/blob/trunk/test-cases/gutenberg/writing-flow/rich-text-formatting.md#tc001) - Bold, Italic, strikethrough buttons
- [x] [TC001](https://github.com/wordpress-mobile/test-cases/blob/trunk/test-cases/gutenberg/writing-flow/rich-text-formatting.md#tc001) - Highlight selected text
- [x] [TC001](https://github.com/wordpress-mobile/test-cases/blob/trunk/test-cases/gutenberg/writing-flow/rich-text-formatting.md#tc001) - Highlight text without selection
- [x] [TC002](https://github.com/wordpress-mobile/test-cases/blob/trunk/test-cases/gutenberg/writing-flow/rich-text-formatting.md#tc002) - Alignment buttons
- [x] [TC003](https://github.com/wordpress-mobile/test-cases/blob/trunk/test-cases/gutenberg/writing-flow/rich-text-formatting.md#tc003) - Alignment Split
- [x] [TC004](https://github.com/wordpress-mobile/test-cases/blob/trunk/test-cases/gutenberg/writing-flow/rich-text-formatting.md#tc004) - Link button works without selection
- [x] [TC005](https://github.com/wordpress-mobile/test-cases/blob/trunk/test-cases/gutenberg/writing-flow/rich-text-formatting.md#tc005) - Link button works with a selected word
- [x] [TC006](https://github.com/wordpress-mobile/test-cases/blob/trunk/test-cases/gutenberg/writing-flow/rich-text-formatting.md#tc006) - Adding a link from a copied URL
- [x] [TC007](https://github.com/wordpress-mobile/test-cases/blob/trunk/test-cases/gutenberg/writing-flow/rich-text-formatting.md#tc007) - Test format detection under the cursor
- [x] [TC008](https://github.com/wordpress-mobile/test-cases/blob/trunk/test-cases/gutenberg/writing-flow/rich-text-formatting.md#tc008) - Test formatting doesn't remove leading or trailing whitespace
- [x] [TC009](https://github.com/wordpress-mobile/test-cases/blob/trunk/test-cases/gutenberg/writing-flow/rich-text-formatting.md#tc009) - Test autocorrection doesn't apply formatting to Heading
- [x] [TC010](https://github.com/wordpress-mobile/test-cases/blob/trunk/test-cases/gutenberg/writing-flow/rich-text-formatting.md#tc010) - Test autocorrection doesn't remove formatting from Heading
##### Splitting and merging
- [x] [TC001](https://github.com/wordpress-mobile/test-cases/blob/trunk/test-cases/gutenberg/writing-flow/splitting-merging.md#tc001) - Merge after writing
  - [x] Paragraph
  - [x] Heading
- [x] [TC002](https://github.com/wordpress-mobile/test-cases/blob/trunk/test-cases/gutenberg/writing-flow/splitting-merging.md#tc002) - Merge after selection
  - [x] Paragraph
  - [x] Heading
- [x] [TC003](https://github.com/wordpress-mobile/test-cases/blob/trunk/test-cases/gutenberg/writing-flow/splitting-merging.md#tc003) - Merge after deleting text
  - [x] Paragraph
  - [x] Heading
- [x] [TC004](https://github.com/wordpress-mobile/test-cases/blob/trunk/test-cases/gutenberg/writing-flow/splitting-merging.md#tc004) - Merge after deleting all
  - [x] Paragraph
  - [x] Heading
- [x] [TC005](https://github.com/wordpress-mobile/test-cases/blob/trunk/test-cases/gutenberg/writing-flow/splitting-merging.md#tc005) - Merge multiple blocks
  - [x] Paragraph
  - [x] Heading
- [x] [TC006](https://github.com/wordpress-mobile/test-cases/blob/trunk/test-cases/gutenberg/writing-flow/splitting-merging.md#tc006) - Splitting/merge list block
#### Undo / Redo - Test Cases
- [x] [TC001](https://github.com/wordpress-mobile/test-cases/blob/trunk/test-cases/gutenberg/writing-flow/undo-redo.md#tc001) - Undo/redo block actions
- [x] [TC002](https://github.com/wordpress-mobile/test-cases/blob/trunk/test-cases/gutenberg/writing-flow/undo-redo.md#tc002) - Undo/redo text
- [x] [TC003](https://github.com/wordpress-mobile/test-cases/blob/trunk/test-cases/gutenberg/writing-flow/undo-redo.md#tc003) - Undo/redo text format

**Comment by kbrown9** [comment]
Everything seemed to work well for me. I encountered a few of the known issues, but I did not notice any new ones.

----------

### 🟢  Writing Flow checklist

Build: https://github.com/wordpress-mobile/WordPress-Android/pull/16742#issuecomment-1155203187
Device: Samsung SM-A135U1 (Android version 11)


#### General
- [x] [TC001](https://github.com/wordpress-mobile/test-cases/blob/trunk/test-cases/gutenberg/writing-flow/copy-paste.md#tc001) - Paste formatted text copied from website
- [x] [TC001](https://github.com/wordpress-mobile/test-cases/blob/trunk/test-cases/gutenberg/writing-flow/multiline-components.md#tc001) - Multiline components
  - [x] Quote block
  - [x] Verse block
  - [x] Preformatted block
  - [ ] Code block (DEV only)
  - [x] Pullquote block
#### Rich Text Format
- [x] [TC001](https://github.com/wordpress-mobile/test-cases/blob/trunk/test-cases/gutenberg/writing-flow/rich-text-formatting.md#tc001) - Bold, Italic, strikethrough buttons
- [x] [TC001](https://github.com/wordpress-mobile/test-cases/blob/trunk/test-cases/gutenberg/writing-flow/rich-text-formatting.md#tc001) - Highlight selected text
- [x] [TC001](https://github.com/wordpress-mobile/test-cases/blob/trunk/test-cases/gutenberg/writing-flow/rich-text-formatting.md#tc001) - Highlight text without selection
- [x] [TC002](https://github.com/wordpress-mobile/test-cases/blob/trunk/test-cases/gutenberg/writing-flow/rich-text-formatting.md#tc002) - Alignment buttons
- [x] [TC003](https://github.com/wordpress-mobile/test-cases/blob/trunk/test-cases/gutenberg/writing-flow/rich-text-formatting.md#tc003) - Alignment Split
- [x] [TC004](https://github.com/wordpress-mobile/test-cases/blob/trunk/test-cases/gutenberg/writing-flow/rich-text-formatting.md#tc004) - Link button works without selection
- [x] [TC005](https://github.com/wordpress-mobile/test-cases/blob/trunk/test-cases/gutenberg/writing-flow/rich-text-formatting.md#tc005) - Link button works with a selected word
- [x] [TC006](https://github.com/wordpress-mobile/test-cases/blob/trunk/test-cases/gutenberg/writing-flow/rich-text-formatting.md#tc006) - Adding a link from a copied URL
- [x] [TC007](https://github.com/wordpress-mobile/test-cases/blob/trunk/test-cases/gutenberg/writing-flow/rich-text-formatting.md#tc007) - Test format detection under the cursor
- [x] [TC008](https://github.com/wordpress-mobile/test-cases/blob/trunk/test-cases/gutenberg/writing-flow/rich-text-formatting.md#tc008) - Test formatting doesn't remove leading or trailing whitespace
- [x] [TC009](https://github.com/wordpress-mobile/test-cases/blob/trunk/test-cases/gutenberg/writing-flow/rich-text-formatting.md#tc009) - Test autocorrection doesn't apply formatting to Heading
- [x] [TC010](https://github.com/wordpress-mobile/test-cases/blob/trunk/test-cases/gutenberg/writing-flow/rich-text-formatting.md#tc010) - Test autocorrection doesn't remove formatting from Heading
##### Splitting and merging
- [x] [TC001](https://github.com/wordpress-mobile/test-cases/blob/trunk/test-cases/gutenberg/writing-flow/splitting-merging.md#tc001) - Merge after writing
  - [x] Paragraph
  - [x] Heading
- [x] [TC002](https://github.com/wordpress-mobile/test-cases/blob/trunk/test-cases/gutenberg/writing-flow/splitting-merging.md#tc002) - Merge after selection
  - [x] Paragraph
  - [x] Heading
- [x] [TC003](https://github.com/wordpress-mobile/test-cases/blob/trunk/test-cases/gutenberg/writing-flow/splitting-merging.md#tc003) - Merge after deleting text
  - [x] Paragraph
  - [x] Heading
- [x] [TC004](https://github.com/wordpress-mobile/test-cases/blob/trunk/test-cases/gutenberg/writing-flow/splitting-merging.md#tc004) - Merge after deleting all
  - [x] Paragraph
  - [x] Heading
- [x] [TC005](https://github.com/wordpress-mobile/test-cases/blob/trunk/test-cases/gutenberg/writing-flow/splitting-merging.md#tc005) - Merge multiple blocks
  - [x] Paragraph
  - [x] Heading
- [x] [TC006](https://github.com/wordpress-mobile/test-cases/blob/trunk/test-cases/gutenberg/writing-flow/splitting-merging.md#tc006) - Splitting/merge list block
#### Undo / Redo - Test Cases
- [x] [TC001](https://github.com/wordpress-mobile/test-cases/blob/trunk/test-cases/gutenberg/writing-flow/undo-redo.md#tc001) - Undo/redo block actions
- [x] [TC002](https://github.com/wordpress-mobile/test-cases/blob/trunk/test-cases/gutenberg/writing-flow/undo-redo.md#tc002) - Undo/redo text
- [x] [TC003](https://github.com/wordpress-mobile/test-cases/blob/trunk/test-cases/gutenberg/writing-flow/undo-redo.md#tc003) - Undo/redo text format

File: packages/rich-text/src/component/index.native.js
Comment (SiobhyB) [inline packages/rich-text/src/component/index.native.js]: It looks like this space was added by mistake.

**Review by SiobhyB — COMMENTED** [review]
@kbrown9, these changes look great to me!

To measure this, I used [Android's dumpys tool](http://android-doc.github.io/preview/testing/performance.html) for measuring performance and followed these steps in an emulator:

* Copied/pasted the content from Automattic's [How We Work](https://automattic.com/how-we-work/) page to a new post. This was in order to get a fair sized post for testing.
* Opened the post and ran `adb shell dumpsys gfxinfo org.wordpress.android.prealpha reset` from Android Studio's terminal, in order to reset any existing data.
* Added three new paragraph blocks to the post, making sure to add multiple lines of content to one. 
* I ran `adb shell dumpsys gfxinfo org.wordpress.android.prealpha` to gather data from the terminal.
* I repeated the above against `trunk` and then with the changes from this PR applied. I was mindful to add similar types of content when testing both the "before" and "after".

Here's an overview of the time it took to render frames at the 50th, 90th, 95th, and 99th percentiles (keeping in mind the desired end goal is to be ≤ 16.67ms in order to [get 60 frames per second](https://www.youtube.com/watch?v=CaMTIgxCSqU&list=PLWz5rJ2EKKc9CBxr3BVjPTPoDPLdPIFCE&index=26)):

-- | Before Fix | After Fix
-- | -- | --
Time to render 50th percentile | 6ms | 5ms
Time to render 90th percentile | 44ms | 26ms
Time to render 95th percentile | 93ms | 42ms
Time to render 99th percentile | 350ms | 121ms

The above indicates improvements of 50%+ for the higher percentiles. Anecdotally, I also felt that things seemed faster when experimenting with adding various blocks and writing content in posts. 😍 

I plan to ask for another set of eyes on this PR in order to double-check that the methodology I used for testing seems sound, so will hold off on approving for now, but I think this seems like it will be a really great improvement!

It'd be a good idea to add this change to the [changelog](https://github.com/WordPress/gutenberg/blob/8d6b3d16a374dc1c1f7c993fcd7c3ff43e8df994/packages/react-native-editor/CHANGELOG.md) in order to make testers aware of it for beta testing too.

File: packages/rich-text/src/component/index.native.js
Comment (kbrown9) [reply packages/rich-text/src/component/index.native.js]: Fixed in commit fae4934.

**Comment by geriux** [comment]
Hey there 👋 these changes look good to me ✅, I manually tested using a Redmi Note 8T and I can see a nice improvement in performance, especially when typing fast, comparing it with the production version.

I did several tests with formatting, switching between blocks, and autocorrecting words.

I think we should add a comment with the code we are adding, just to have a reference of the problem is solving. What do you think @kbrown9?

**Comment by kbrown9** [comment]
> Hey there 👋 these changes look good to me ✅, I manually tested using a Redmi Note 8T and I can see a nice improvement in performance, especially when typing fast, comparing it with the production version.
> 
> I did several tests with formatting, switching between blocks, and autocorrecting words.
> 
> I think we should add a comment with the code we are adding, just to have a reference of the problem is solving. What do you think @kbrown9?

Thanks for the tests and review! I agree that a comment with the code makes sense, and I'll also add this change to the changelog.

**Review by SiobhyB — APPROVED** [review]
Thanks for addressing the comments so quickly @kbrown9 and for reviewing @geriux. Approving, as this looks good to me :D 🎉

**Pull request merged**
